src/hotspot/os/linux/os_linux.cpp

```diff
@@ -4746,19 +4746,46 @@ int os::active_processor_count() {
   return active_cpus;
 }
 
+static volatile int warn_invalid_processor_id = 1;
+
+static bool should_warn_invalid_processor_id() {
+  if (os::processor_count() == 1) {
+    // Don't warn if we only have one processor
+    return false;
+  }
+
+  if (Atomic::load(&warn_invalid_processor_id) == 0 ||
+      Atomic::xchg(&warn_invalid_processor_id, 0) == 0) {
+    // Don't warn more than once
+    return false;
+  }
+
+  return true;
+}
+
 uint os::processor_id() {
   const int id = Linux::sched_getcpu();
 
-#ifndef PRODUCT
-  if (UseDebuggerErgo1 && id >= _processor_count) {
-    // Some debuggers limit the processor count without limiting
-    // the returned processor ids. Fake the processor id.
-    return 0;
+  if (id >= 0 && id < processor_count()) {
```

Comment:
Do we really need to check if the returned processor ID is negative? That seems a whole new level of environment screwup to me.

Reply:
I'm thinking we should make this safe to call in all cases. God knows what a broken environment might return.

Reply:
After some discussions, we agreed to not to check for negative processor ids.